### PR TITLE
CP-52821: use Mtime in Xapi_periodic_scheduler

### PR DESCRIPTION
**Pull request by edwintorok**
Target: feature/perf

Note that this conflicts with https://github.com/xapi-project/xen-api/pull/6155/commits/7e9310f110df2613229bbb236334aad706209a31?diff=unified&w=1, but is a pre-requisite of https://github.com/xapi-project/xen-api/pull/6126

### CODE REVIEW

**Comment by robhoes** [comment]
Isn't it better to rebase on master already and build on top of the scheduler changes there? Those rebase changes would have to be done later anyway, but then end up in a complicated and hard to review merge commit.

File: ocaml/xapi/xapi_periodic_scheduler.ml [hunk 1]
```diff
@@ -88,7 +88,7 @@ let loop () =
       else
         let next = with_lock lock (fun () -> Ipq.maximum queue) in
         let now = Mtime_clock.now () in
-        if next.Ipq.time < now then (
+        if Mtime.compare next.Ipq.time now < 0 then (
```
Comment (psafont) [inline ocaml/xapi/xapi_periodic_scheduler.ml:91]:
```suggestion
        if Mtime.is_earlier next.Ipq.time ~than time then (
```

**Comment by psafont** [comment]
There's a further improvement here: replace all `Mtime_clock.now ()` with `Mtime_clock.elapsed()` to remove the overflow treatment because the latter returns a span isntead of a timestamp

File: ocaml/xapi/xapi_periodic_scheduler.ml (hunk 1 above)
Comment (edwintorok) [reply ocaml/xapi/xapi_periodic_scheduler.ml]: almost, but had to use `now` instead of `time`, I pushed that as a separate commit.

**Comment by edwintorok** [comment]
> Isn't it better to rebase on master already and build on top of the scheduler changes there? Those rebase changes would have to be done later anyway, but then end up in a complicated and hard to review merge commit.

I'll revise this PR on top of master once Frediano's PR is merged.

**Comment by freddy77** [comment]
> There's a further improvement here: replace all `Mtime_clock.now ()` with `Mtime_clock.elapsed()` to remove the overflow treatment because the latter returns a span instead of a timestamp

It's not clear what overflows you are talking about. Internally the timer is a counter since boot, that will overflow the `int64` (although the numbers should be unsigned) in about 290 years. I personally find the code with `Mtime_clock.elapsed()` less readable having to deal with differences of differences (this span is less or more than this other span) instead of just comparing time (the deadline cross current time).

**Comment by psafont** [comment]
> It's not clear what overflows you are talking about. 

One function returns an option in case of overflow, the other does not. Because, as you point out, the overflow won't happen in practise, I prefer to use the latter rather than the former.

**Review by psafont — APPROVED** [review]

**Comment by edwintorok** [comment]
Moved this to draft until the referenced PR is merged to master as this will need a rebase. https://github.com/xapi-project/xen-api/pull/6155

**Comment by edwintorok** [comment]
Rebased on latest feature/perf (which in turn has been updated to include latest master).
Target is still feature/perf because there are more changes coming, and this needs a bit wider testing.

E.g. previously there was a +0.5 in the periodic scheduler, but now we just use Clock.Timer.remaining, so there could be some subtle differences in behaviour.

**Comment by edwintorok** [comment]
Rebased on latest `feature/perf` which incorporated the latest changes to the scheduler from master.

**Review by freddy77 — APPROVED** [review]

**Comment by freddy77** [comment]
"Minor" hidden change. Due to internal implementation moving from `Mtime_clock.now` to `Mtime_clock.elapsed` on Linux move the timer from `CLOCK_MONOTONIC` to `CLOCK_BOOTTIME`. They are both monotonic but the second takes into account when the machine is suspended. But in our case machine should not be suspended so it should be the same.

**Comment by edwintorok** [comment]
> move the timer from `CLOCK_MONOTONIC` to `CLOCK_BOOTTIME`

For something like 'run this task every 24h' that actually seems like the correct behaviour (even if the machine was asleep we'd want it to run the task if it wakes up sooner than that), so the periodic tasks should probably use BOOTTIME (i.e. `Mtime_clock.elapsed`), and oneshot tasks monotonic (i.e. `Mtime_clock.now`).

But as you said we don't actually suspend Dom0, so we don't need to complicate the scheduler to make that distinction.

**Pull request merged**